### PR TITLE
Updater

### PR DESCRIPTION
**Pull request by gabriel**
Updater in go (service). Includes commands and API for manually triggering updates from clients (or checking updates).

It's currently 100% automatic, but we'll want to add config to allow users to preview updates.

Remaining things on my list, though should go pretty quick:

- Notify the app to restart when an update has been applied
- Service API call. I am using this static endpoint currently for testing: https://keybase-app.s3.amazonaws.com/update-1.0.5-6.json?osname=darwin&runmode=prod&version=1.0.5-6
- Handle error if users made application owned by non-admin

### CODE REVIEW

**Comment by maxtaco** [comment]
Can you make your JSON reply match [this format](https://keybase.io/_/api/1.0/user/lookup.json?username=max&fields=basics)?  This way we can use the same API Go client libraries.

File: go/install/update.go [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+	return hex.EncodeToString(hash.Sum(result)), nil
+}
+
+func (u *Updater) pathForFilename(filename string) string {
+	return path.Join(os.TempDir(), "KeybaseUpdates", filename)
+}
+
+func isFileScheme(url string) (string, bool) {
+	scheme := "file://"
+	if strings.HasPrefix(url, scheme) {
```
Comment (maxtaco) [inline go/install/update.go]: The [Golang URL library](https://golang.org/pkg/net/url/#URL) has these features.

File: go/install/update.go [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+	}
+	etag := ""
+	if _, err = os.Stat(fpath); err == nil {
+		etag, err = computeEtag(fpath)
+		if err != nil {
+			return
+		}
+	}
+
+	req, _ := http.NewRequest("GET", url, nil)
```
Comment (maxtaco) [inline go/install/update.go]: We should use the Keybase API to do these calls.  See libkb/api.go. There's an `API` object dangling off of `G`.

File: go/install/update.go [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+		cached = true
+		return
+	}
+	if resp.StatusCode != http.StatusOK {
+		err = fmt.Errorf("Responded with %s", resp.Status)
+		return
+	}
+
+	savePath := fmt.Sprintf("%s.download", fpath)
+	file, err := os.OpenFile(savePath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, libkb.PermFile)
```
Comment (maxtaco) [inline go/install/update.go]: Will this work for remote URLs?

File: go/install/update.go [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+		if err != nil {
+			return
+		}
+	}
+	u.G().Log.Info("Moving %s to %s", path.Base(savePath), path.Base(fpath))
+	err = os.Rename(savePath, fpath)
+	return
+}
+
+func (u *Updater) unpack(filename string) (string, error) {
```
Comment (maxtaco) [inline go/install/update.go]: What's our story going to be for signing and verifying new downloads?  All within the DMG?  We have the opportunity to check here too, and maybe in a way that doesn't depend on the Root CAs or Apple.

File: go/tools/release/github/releases.go [hunk 5]
```diff
@@ -59,6 +60,14 @@ func Releases(user, repo, token string) ([]Release, error) {
 	return releases, nil
 }
 
+func LatestRelease(user, repo, token string) (*Release, error) {
+	if token != "" {
+		token = "?access_token=" + token
```
Comment (maxtaco) [inline go/tools/release/github/releases.go]: Use the [Go API](https://golang.org/pkg/net/url/#URL) to piece together this URL.

**Comment by maxtaco** [comment]
See failing CI checks...

**Comment by maxtaco** [comment]
What's the story for having both the GitHub and the self-hosted sources for updates?

File: go/install/update.go (hunk 1 above)
Comment (gabriel) [inline go/install/update.go]: Fixed

File: go/install/update.go (hunk 2 above)
Comment (gabriel) [inline go/install/update.go]: Fixed

File: go/install/update.go (hunk 3 above)
Comment (gabriel) [inline go/install/update.go]: yeah been testing with remote and local urls

**Comment by gabriel** [comment]
The updater now points to the API endpoint: 

```
/update.json?version=1.2.3-400&os_name=darwin&run_mode=prod&channel=nightly
```

Was thinking channel could be like chrome update channels (release, beta, dev, nightly, etc)

The response should look something like:
```
{
	"status": {
    "code": 0,
    "name": "OK"
  },
	"update": {
		"version":     "1.0.5-7",
		"name":        "Keybase",
		"description": "Bug fixes",
		"asset": {
			"name":        "Keybase-1.0.5-7.zip",
			"url": "htp://..../Keybase-1.0.5-7.zip"
		}
	}
}
```

we can change the format (it's defined in the protocol) if you have more feedback there.

File: go/install/update.go (hunk 4 above)
Comment (gabriel) [inline go/install/update.go]: Not sure... should the update include a signature in the asset that the client checks?

**Comment by gabriel** [comment]
The asset could have a kbfs location and/or sig?

```json
"asset": {
  "name":        "Keybase-1.0.5-7.zip",
  "url": "https://..../Keybase-1.0.5-7.zip",
  "sig": "...",
  "kbfs": "/keybase/public/keybase/Keybase-1.0.5-7.zip"
}
```

**Comment by gabriel** [comment]
I think it's useful to keep around the other sources (github, custom url). They are helpful to me for testing.

File: go/install/sources/github.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+		return
+	}
+	g.G().Log.Debug("Found Github release: %#v", ghRelease)
+
+	uver := ghRelease.TagName[1:]
+
+	// Find the asset matching asset
+	var asset *keybase1.Asset
+	for _, a := range ghRelease.Assets {
+		if strings.HasPrefix(a.Name, "Keybase-") && strings.HasSuffix(a.Name, fmt.Sprintf("-%s.zip", runtime.GOOS)) {
```
Comment (maxtaco) [inline go/install/sources/github.go]: will it be upper case?

File: go/install/sources/keybase.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+	var res updateResponse
+	err = k.G().API.GetDecode(libkb.APIArg{
+		Endpoint: "update.json",
+		Args:     APIArgs,
+	}, res)
+	if err != nil {
+		return
+	}
+	if res.Status.Code != libkb.SCOk {
```
Comment (maxtaco) [inline go/install/sources/keybase.go]: The library already checks this, unless you've found a bug....  (see `checkStatus`...)

File: go/install/update.go (hunk 3 above)
Comment (maxtaco) [inline go/install/update.go]: whoops, this is just for the save, my bad...

File: go/install/update.go (hunk 4 above)
Comment (maxtaco) [inline go/install/update.go]: that's a good idea.  we can make a ticket for that

**Comment by maxtaco** [comment]
Looks good, but two things: I had another question, and see Windows C/I failure. Might have to break it into _nix and _windows file for the FS-walk permissions-checking stuff.

**Comment by gabriel** [comment]
Totally borked this branch... Moved to https://github.com/keybase/client/pull/1510

**Pull request closed without merging**